pkg/runctl/controller.go

```diff
@@ -304,6 +298,9 @@ func (c *Controller) syncHandler(key string) error {
 	// If not, try to add finalizer if missing
 	if pipelineRun.HasDeletionTimestamp() {
 		runManager := c.createRunManager(pipelineRun)
+		if pipelineRun.GetStatus().State == api.StateFinished {
```

Comment:
We should  always run `runManager.Cleanup` before remove the finalizer.

Reply:
This is fixed with 30010ca31e4032a4d90664e79633a8432e35b910.
We only set state to StateFinished in the finish() function.
Cleanup is called before each of the calls of finish().

Reply:
If we don't add this statement, `c.finish()` will then be called, right? I think we don't need it. Am I missing something?

Reply:
We need it, otherwise we would potenitially change a Result. e.g. Result is set to Success but finalizer deletion failed.
If then a deletion occurs before next resync a Result Success would be changed to Result Deleted.